### DISCUSSION TITLE
Help troubleshooting SLEAP centered-instance model predictions

### DISCUSSION THREAD

**Discussion by Taschimoz** [talmolab/sleap#2748] [Help!] [answered]
The centroid/anchor point, TTI, appears to be detected accurately, but the other body landmarks are being predicted incorrectly. Specifically, the upper-body points and tail points seem compressed or clustered near the TTI instead of spreading across the full mouse body and tail. Additionally, not all the nodes or points are appearing which I assume could be due to a lack of frames or a error in training. 

I initially thought this might be due to too few labeled frames, since I have around 300 labeled frames so far. However, the centroid predictions look good, and the training error is very low, around 1 × 10^-5, so I suspect the issue may be related to the centered-instance configuration, anchor choice, or label distribution rather than just the number of labels.

I have tried adjusting the centered-instance configuration, including:

Input Scaling: 1.00
Crop Size: 512
Max Stride: 64
Anchor Part: TTI
Training from scratch for centered instance rather than fine-tuning
or 
Input Scaling: 0.75
Crop Size: 512
Max Stride: 64
Anchor Part : TTI

Even after these changes, the predictions still show the same issue: the landmarks remain clustered near TTI, while the centroid itself is accurate. I have also ensured that the red crop box covers the entire animal and the blue box covers the TTI. 

Would you be able to help me troubleshoot this or point me toward the best next steps? I’m wondering in particular whether TTI is a poor anchor point for the centered-instance model, whether I should add a more central body node, whether the receptive field/model architecture should be changed, or whether I simply need to add more targeted labeled frames with extended tails and varied body postures.

Thank you so much!
<img width="1152" height="1536" alt="image" src="https://github.com/user-attachments/assets/12fc09e8-b248-4abe-9ae3-b334b10ab4f8" />
<img width="1152" height="1536" alt="image" src="https://github.com/user-attachments/assets/5ae0f2c5-ba38-41dc-aa5f-b5ab3d407243" />

**Comment by tom21100227** [comment]
Hi @Taschimoz , 

This is weird, the prediction output you're showing is clearly a crop size issue but your config is correct. My guess is either 1) your training data is labelled like that which is unlikely, or 2) when running inference on new videos you stick with the original model instead of the new model. 

To troubleshoot that can you walk us through how you're running inference on new videos? Can you find where the newly trained model is located, make sure you're selected the `training_config.yaml`/`json` from the newly trained model. I would also try bypass the GUI and specify the exact model with `sleap-nn` for now to narrow it down, [tutorials on how to do that is here.](https://nn.sleap.ai/latest/guides/inference/)

Best,
Tom

**Comment by Taschimoz** [comment]
Hi Tom,

My training data and the inferences that the model are outproduced are pasted below. I stick with the new model when I run inference. Is there perhaps a scaling issue because it seems like the instances predicted are compressed but the validation image (pasted below) seems fine. The way Im running inference on the new videos is selecting the training_config.yaml of the most recent model and then running inference. I also tried bypassing the GUI and specifying the exact model and the predicted frames remain the same. Images pasted below for reference
<img width="2160" height="3840" alt="D681A934-7195-4CA2-9BA6-A0D4078D2454_0AEBE1A9-D72E-4F06-85A5-62BF5CA810CA" src="https://github.com/user-attachments/assets/3f13ffbc-abd5-4fb9-b665-58926ec0e761" />
<img width="3024" height="4032" alt="IMG_5879" src="https://github.com/user-attachments/assets/1c2045d0-6f7b-444d-9042-da9fe649e489" />
<img width="3024" height="4032" alt="IMG_5881" src="https://github.com/user-attachments/assets/0c8abcb4-e73a-45ac-8c0c-fab03ebfc46f" />
<img width="3024" height="4032" alt="IMG_5882" src="https://github.com/user-attachments/assets/f9d200c4-dac1-44d3-ac89-7322327746c1" />
<img width="3024" height="4032" alt="IMG_5884" src="https://github.com/user-attachments/assets/e7543eae-908c-4e3f-9d07-de8a00a480dd" />
<img width="3024" height="4032" alt="IMG_5885" src="https://github.com/user-attachments/assets/66d7c674-4213-4f93-b25e-0f337b5035d4" />
<img width="3024" height="4032" alt="IMG_5886" src="https://github.com/user-attachments/assets/d1f9f0fb-cef3-4324-b0a1-97f6f0dc05a7" />

**Comment by tom21100227** [comment]
Hi @Taschimoz , 

A few things I would like you to try: 
1. Set crop size to auto and rerun
2. upload your slp file and training config yaml [here](https://docs.google.com/forms/d/e/1FAIpQLSfWBwcuZH8lxGyFabMAODLjSUE4YrSCYbrrnYvXCpgBqZpcJA/viewform) so we can take a look
3. if possible, upgrade your sleap version to 1.6.3. [Do a fresh installation](https://docs.sleap.ai/latest/installation/#install-sleap).

**Comment by Taschimoz** [comment]
Just to confirm, you want me to rerun training with crop size to auto. It seems like I am unable to change the inference configurations. I haev uploaded the slp file and will get back to you in regards to upgrading my sleap version. Thanks!

**Reply by Taschimoz** [reply]
I upgraded my sleap version to 1.6.3 and ran training and inference with crop size on auto and the issue still remains.

**Reply by tom21100227** [reply]
@Taschimoz Hi Kerry, 

Sorry about the delay, I was out sick last week, can you give tom@talmolab.org permission to the drive link you uploaded?

**Comment by tom21100227** [comment]
Hi @Taschimoz , 

Seems like a config mismatch from a legacy import error, in your centered instance model (the one without `(1)`), edit the yaml on the `preprocessing` line. 

```yaml
  max_height: null   # was 480
  max_width: null    # was 640
```

In the meantime can you upload the 3 videos attached to the slp file? I want to see if we can recreate your error on our end. 

Tom

**Comment by Taschimoz** [comment]
Here is the google drive with the slp files: https://drive.google.com/file/d/1m_LEJ46NzE7HxnIm0DLc8JkDEOTi8Cjc/view?usp=sharing in case you can access and then, due to cluster maintenance, I will upload the three videos on thursday as soon as it is back up.

**Reply by tom21100227** [reply]
when you got time, try and set max height and width to null and retrain, I believe that should fix the issue, but if not I will need the videos and see if I can recreate this error.

**Comment by Taschimoz** [comment]
It got fixed. Thanks so so much!

**Comment by tom21100227** [comment]
Hi @Taschimoz , glad to hear it's working! 

Just to prevent other users from getting the same bug that you're facing, can you tell us what steps did you do to create that training config? Was it from a legacy (<= 1.4.x version) sleap? 

Best,
Tom

**Comment by Taschimoz** [comment]
Hi Tom,

Thanks for following up! To clarify the steps that led to this issue:

1. I ran a training session once to generate the initial training config file or just edited my previously ran file. 
2. After the run, I manually edited the config file, setting `max_height` and `max_width` to `null`.

This was not done on a legacy (≤ 1.4.x) version of SLEAP — I was using a more recent version.

Hope that helps narrow down the source of the bug. Happy to provide any additional details if needed.

**Comment by tom21100227** [comment]
https://github.com/user-attachments/assets/88b8d996-be40-495c-ad3b-77f1f2cb21c8

@Taschimoz trying to recreate this bug, does this looks like what you did to create that config?

**Reply by Taschimoz** [reply]
Yes

**Comment by tom21100227** [comment, marked as answer]
Thanks @Taschimoz — your detail let us pin this down completely, and you were right that it wasn't a legacy import.

**Root cause.** When training a centered-instance (top-down) model, SLEAP was auto-filling `max_height`/`max_width` in the saved `training_config.yaml` from your *training* video's dimensions. A centered-instance model works on fixed-size crops around the anchor (`crop_size`), so it shouldn't carry a full-frame `max_height`/`max_width` at all. At inference those values make SLEAP resize each frame to that size **before** cropping — so on videos at a different resolution than your training video, the animal gets rescaled inside the crop and the body nodes collapse toward the anchor. Your validation frames looked fine because they're the same resolution as training (the resize was a no-op there). Setting both to `null`, as you did, crops at native resolution → fixed.

We reproduced it end-to-end: predictions are correct only when inference resolution matches training, and collapse hardest when they differ (down to ~7% of the correct node spread on a 3× upscaled canvas), while `null` stays correct at every resolution.

**Fix.** talmolab/sleap-nn#679 stops `max_height`/`max_width` from being set on crop-based models in both places they could come from (the training setup and the legacy-config importer). Until that ships, the workaround is exactly what you did: set `max_height: null` and `max_width: null` in the centered-instance `training_config.yaml`.

Thanks again for the thorough back-and-forth and the sample files — closing this out.
